### ISSUE TITLE
[Question] Examples with PHP / Symfony

### ISSUE THREAD

**Issue by ToshY** [lexiforest/curl-impersonate#93]
I've been trying to setup `curl-impersonate` with Symfony's [CurlHttpClient](https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/HttpClient/CurlHttpClient.php) while using Docker, but I'm failing to get it to work. The binaries work fine on there own so it seems related to the extension/how to use it.

I've tried according to the original [docs](https://github.com/lwthiker/curl-impersonate/blob/main/docs/03_LIBCURL_IMPERSONATE_PHP.md#on-linux) to install, but without any help. Also tried a [comment ](https://git.lolcat.ca/lolcat/4get/src/branch/master/docs/configure.md#cloudflare-bypass-tls-check) I saw in the original repo, but also didn't work.

The question is if there are any users of `curl-impersonate` that got this working in PHP/Symfony to have the curl client use `curl-impersonate` internally. Any bit of imformation would be appreciated.

---

`Dockerfile` (part for installing curl-impersonate)
```Dockerfile
RUN <<EOT bash
  set -ex
  curl -L --retry 3 --retry-connrefused --retry-delay 2 --fail-with-body -o /tmp/curl-impersonate.tar.gz https://github.com/lexiforest/curl-impersonate/releases/download/v0.8.2/curl-impersonate-v0.8.2.x86_64-linux-gnu.tar.gz
  tar -xzf /tmp/curl-impersonate.tar.gz -C /usr/local/bin --no-same-owner
  curl -L --retry 3 --retry-connrefused --retry-delay 2 --fail-with-body -o /tmp/libcurl-impersonate.tar.gz https://github.com/lexiforest/curl-impersonate/releases/download/v0.8.2/libcurl-impersonate-v0.8.2.x86_64-linux-gnu.tar.gz
  tar -xzf /tmp/libcurl-impersonate.tar.gz -C /tmp --no-same-owner
  mv /tmp/libcurl-impersonate-* /usr/lib/x86_64-linux-gnu/
  cp /usr/lib/x86_64-linux-gnu/libcurl-impersonate-chrome.so.4.8.0 /usr/lib/x86_64-linux-gnu/libcurl.so.4.8.0
  rm -f /tmp/*
  ldconfig
EOT
```

---

**Test binary**
```shell
curl_chrome131 -s -X POST https://kitsu.app/api/graphql -H "Content-Type: application/json" -d '{"query": "query { findAnimeById(id: 1) { id } }"}'
```

**Respone**
```json
{
   "data":{
      "findAnimeById":{
         "id":"1"
      }
   }
}
```

---

**Test application**
Symfony POST request.

```php
$result = $this->httpClient->request(
    'POST',
    'https://kitsu.app/api/graphql',
    [
        'json' => ['query' => 'query { findAnimeById(id: 1) { id } }"}'],
    ]
);
```

**Response**
Returns `403` due to CloudFlare (likely fingerprint?)
![image](https://github.com/user-attachments/assets/c87a0ed0-82f7-4463-a8e4-e66c73d134a2)

**Standalone cURL**

`php snippet.php`
```php
<?php
$url = 'https://kitsu.app/api/graphql';
$query = '{"query": "query { findAnimeById(id: 1) { id } }"}';

$ch = curl_init($url);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_POST, true);
curl_setopt($ch, CURLOPT_HTTPHEADER, [
    'Content-Type: application/json',
]);
curl_setopt($ch, CURLOPT_POSTFIELDS, $query);

$response = curl_exec($ch);
if ($response === false) {
    echo 'cURL Error: ' . curl_error($ch);
} else {
    echo "Response: " . $response;
}

curl_close($ch);
```

**Response**

Returns `403` again.

```text
docker compose exec phpfpm php snippet.php
php: /lib/x86_64-linux-gnu/libcurl.so.4: no version information available (required by php)
Response: <!DOCTYPE html><html lang="en-US"><head><title>Just a moment...</title><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><meta name="robots" content="noindex,nofollow"><meta name="viewport" content="width=device-width,initial-scale=1"><style>*
```

---

`php -r 'print_r(curl_version());'` (in container)
```text
docker compose exec phpfpm php -r 'print_r(curl_version());'
php: /lib/x86_64-linux-gnu/libcurl.so.4: no version information available (required by php)
Array
(
    [version_number] => 526080
    [age] => 10
    [features] => 1437139613
    [feature_list] => Array
        (
            [AsynchDNS] => 1
            [CharConv] =>
            [Debug] =>
            [GSS-Negotiate] =>
            [IDN] =>
            [IPv6] => 1
            [krb4] =>
            [Largefile] => 1
            [libz] => 1
            [NTLM] => 1
            [NTLMWB] =>
            [SPNEGO] =>
            [SSL] => 1
            [SSPI] =>
            [TLS-SRP] =>
            [HTTP2] => 1
            [GSSAPI] =>
            [KERBEROS5] =>
            [UNIX_SOCKETS] => 1
            [PSL] =>
            [HTTPS_PROXY] => 1
            [MULTI_SSL] =>
            [BROTLI] => 1
            [ALTSVC] => 1
            [HTTP3] =>
            [UNICODE] =>
            [ZSTD] => 1
            [HSTS] => 1
            [GSASL] =>
        )

    [ssl_version_number] => 0
    [version] => 8.7.0-DEV
    [host] => x86_64-pc-linux-gnu
    [ssl_version] => BoringSSL
    [libz_version] => 1.3
    [protocols] => Array
        (
            [0] => dict
            [1] => file
            [2] => ftp
            [3] => ftps
            [4] => gopher
            [5] => gophers
            [6] => http
            [7] => https
            [8] => imap
            [9] => imaps
            [10] => mqtt
            [11] => pop3
            [12] => pop3s
            [13] => rtsp
            [14] => smb
            [15] => smbs
            [16] => smtp
            [17] => smtps
            [18] => telnet
            [19] => tftp
            [20] => ws
            [21] => wss
        )

    [ares] =>
    [ares_num] => 0
    [libidn] =>
    [iconv_ver_num] => 0
    [libssh_version] =>
    [brotli_ver_num] => 16781312
    [brotli_version] => 1.1.0
)
```

**Comment by lexiforest** [comment]
Did you set the `CURL_IMPERSONATE` environment variable?

**Comment by ToshY** [comment]
> Did you set the `CURL_IMPERSONATE` environment variable?

Well ... no 🤦Thanks for that.

Still, after adding it to the Dockerfile though (with `ENV CURL_IMPERSONATE='chrome131'`) it works for the basic snippet example above:

```php
<?php
$url = 'https://kitsu.app/api/graphql';
$query = '{"query": "query { findAnimeById(id: 1) { id } }"}';

$ch = curl_init($url);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_POST, true);
curl_setopt($ch, CURLOPT_HTTPHEADER, [
    'Content-Type: application/json',
]);
curl_setopt($ch, CURLOPT_POSTFIELDS, $query);

$response = curl_exec($ch);
if ($response === false) {
    echo 'cURL Error: ' . curl_error($ch);
} else {
    echo "Response: " . $response;
}

curl_close($ch);
```

_Result_

(not sure about the version warning here though)

```text
php: /lib/x86_64-linux-gnu/libcurl.so.4: no version information available (required by php)
Response: {"data":{"findAnimeById":{"id":"1"}}}
```

**But** not for existing Symfony HTTP clients calls

```php
$result = $this->httpClient->request(
    'POST',
    'https://kitsu.app/api/graphql',
    [
        'json' => ['query' => 'query { findAnimeById(id: 1) { id } }"}'],
    ]
);

dump($result->getContent());
```

_Result_

![image](https://github.com/user-attachments/assets/f5dd997a-3dd7-4125-84c8-8311f9daeb57)

---

It seems to follow the pattern of failing from `AsyncResponse`.

![image](https://github.com/user-attachments/assets/bcfde1da-f896-4078-905a-ea3372f23691)

I was not sure, but it seems related to getting the (compressed) data (by looking on what [`inflate_add`](https://www.php.net/manual/en/function.inflate-add.php#refsect1-function.inflate-add-returnvalues) does).

---

After trying several things, I've found that when adding the header `Accept-Encoding: identity`, it will return uncompressed response, and it works without problems.

Now adding that header is just a workaround and I can't even convince myself it's a good one at that.

@lexiforest If you have ideas on why this occurs, or have additional ideas/suggestion please let me know 👍

**Comment by lexiforest** [comment]
By default, curl-impersonate add the browser default headers, which include "Accept-Encoding: gzip...". I guess the client you are using conflict with the implictly added headers. You may want to disable that behavior with "CURL_IMPERSONATE_HEADERS", see the docs on README for details.

**Comment by ToshY** [comment]
@lexiforest Thanks for the response.

Let me first start of by saying that I've finally have the given example working with the default HTTP client, without additional `Accept-Encoding` headers or the need to use `CURL_IMPERSONATE_HEADERS=no`.

This now works:
```php
$result = $this->httpClient->request(
    'POST',
    'https://kitsu.app/api/graphql',
    [
        'json' => ['query' => 'query { findAnimeById(id: 1) { id } }"}'],
    ]
);

dump($result->getContent());
```

With the following in the Dockerfile:

```Dockerfile
ENV CURL_IMPERSONATE=chrome131

COPY --from=ghcr.io/tarampampam/curl:8.11.1 /bin/curl /bin/curl

RUN <<EOT bash
  set -ex
  curl -L --retry 3 --retry-connrefused --retry-delay 2 --fail-with-body -o /tmp/curl-impersonate.tar.gz https://github.com/lexiforest/curl-impersonate/releases/download/v0.8.2/curl-impersonate-v0.8.2.x86_64-linux-gnu.tar.gz
  tar -xzf /tmp/curl-impersonate.tar.gz -C /usr/local/bin --no-same-owner
  curl -L --retry 3 --retry-connrefused --retry-delay 2 --fail-with-body -o /tmp/libcurl-impersonate.tar.gz https://github.com/lexiforest/curl-impersonate/releases/download/v0.8.2/libcurl-impersonate-v0.8.2.x86_64-linux-gnu.tar.gz
  tar -xzf /tmp/libcurl-impersonate.tar.gz -C /tmp --no-same-owner
  mv /tmp/libcurl-impersonate-* /usr/lib/x86_64-linux-gnu/
  rm -f /tmp/*
  ldconfig
EOT

ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libcurl-impersonate-chrome.so.4.8.0
```

---

After the above was working, I've found myself clicking through my application, and finding that calls that were made internally through other services, like [Meilisearch](https://github.com/meilisearch/meilisearch), were now also failing with `inflate_add(): data error` (from PSR stream).

![image](https://github.com/user-attachments/assets/39220641-4c37-4e97-a259-478d1281b61d)

> I guess the client you are using conflicts with the implictly added headers

After further debugging the [`CurlHttpClient`](https://github.com/symfony/http-client/blob/3f4090d52c4e098756491cfda5540dd0fc081177/CurlHttpClient.php#L213) and [`CurlResponse`](https://github.com/symfony/http-client/blob/3f4090d52c4e098756491cfda5540dd0fc081177/Response/CurlResponse.php#L121) that Symfony uses, I agree with you that there is something causing conflict, and yes it might be related to headers but could also be a bug/flaw in the inflate logic of the response chunks. I will not further bother you with these details and create a bug report in the Symfony repo instead for this to see how that works out. I will report back when I've figured it out over there.

Sidenote: I've created a temporary patch for myself for this for the [`CurlHttpClient`](https://github.com/symfony/http-client/blob/3f4090d52c4e098756491cfda5540dd0fc081177/CurlHttpClient.php#L213).
```patch
--- CurlHttpClient.php	2025-01-09 21:57:43.446365726 +0000
+++ CurlHttpClient.php	2025-01-09 21:57:20.102366512 +0000
@@ -211,6 +211,7 @@

         if (\extension_loaded('zlib') && !isset($options['normalized_headers']['accept-encoding'])) {
             $options['headers'][] = 'Accept-Encoding: gzip'; // Expose only one encoding, some servers mess up when more are provided
+            $options['normalized_headers']['accept-encoding'] = 'gzip';
         }
         $body = $options['body'];
```

**Comment by ToshY** [comment]
@lexiforest After discussuing this in https://github.com/symfony/symfony/issues/59551, the conclusion seems to be that because the patched libcurl _always_ decompresses the response by using `CURLOPT_ACCEPT_ENCODING` (seen [here](https://github.com/lexiforest/curl-impersonate/blob/6a2e69a001dc1c1ebe9b95a025f4576b3e5aa51f/chrome/patches/curl-impersonate.patch#L674)), this caused errors in a Symfony application as this expects uncompressed response data in order for Symfony to handle the decompression itself (for better error handling).

So my question is now, is it possible to add something like another environment, similar to `CURL_IMPERSONATE_HEADERS=no` (possibly `CURL_IMPERSONATE_DECOMPRESS=no`) to disable decompression for the patched libcurl? This would make it a lot more user friendly to use with atleast Symfony (and likely other PHP frameworks as well).

**Comment by lexiforest** [comment]
The header is part of the impersonation suite. If you don't want the headers to be added, simply set `CURL_IMPERSONATE_HEADERS=no`, and then bring you own headers.

**Comment by ToshY** [comment]
> The header is part of the impersonation suite. If you don't want the headers to be added, simply set `CURL_IMPERSONATE_HEADERS=no`, and then bring you own headers.

I understand that the headers are part of the suite, but I'm trying to convey that's not the issue. Symfony will be able to override headers without problems, with or without `CURL_IMPERSONATE_HEADERS`. It's about this patched libcurl decompressing the response automatically and passing the already decompressed response back to Symfony (which does not expect it to be decompressed already).

Does the logic in [`curl-impersonate.patch#L674`](https://github.com/lexiforest/curl-impersonate/blob/6a2e69a001dc1c1ebe9b95a025f4576b3e5aa51f/chrome/patches/curl-impersonate.patch#L674) not show that it **always** set [`CURLOPT_ACCEPT_ENCODING: ""`](https://curl.se/libcurl/c/CURLOPT_ACCEPT_ENCODING.html) leading to decompression of the response when a `Accept-Encoding` request header was given?

TLDR;

I like to have the ability to get the compressed response from the patched libcurl, not decompressed. 

And to achieve this, I suggested an addtional environment variable (like for example `CURL_IMPERSONATE_DECOMPRESS=no`) with an extra if-check at [`curl-impersonate.patch#L674`](https://github.com/lexiforest/curl-impersonate/blob/6a2e69a001dc1c1ebe9b95a025f4576b3e5aa51f/chrome/patches/curl-impersonate.patch#L674).

**Comment by lexiforest** [comment]
You made a good point here, I was even unaware of this line, it was there long before the fork, and I honestly thought it's the default behavior of curl. I will think of a solution, but it may not be the environment variable way you asked here.

Thanks for digging this out.

**Comment by ToshY** [comment]
> You made a good point here, I was even unaware of this line, it was there long before the fork, and I honestly thought it's the default behavior of curl. I will think of a solution, but it may not be the environment variable way you asked here.
> 
> Thanks for digging this out.

Glad we are on the same page. 

Regarding the possible solution: that's understandable. I've made this suggestion because just like `CURL_IMPERSONATE_HEADERS` it can be easily (un)set, with or without Docker, and makes it very easy to use as instant replacement for existing `libcurl`, which also covers my use case (like denoted in the docs section ["Using CURL_IMPERSONATE env var"](https://github.com/lexiforest/curl-impersonate/blob/main/docs/04_libcurl.md#using-curl_impersonate-env-var)). As frameworks handle most logic for setting curl opts without the user having direct control over this, an additional environment variable seems like a relatively small (non-breaking) change to have.

I'm interested to see what solution(s) you can think of 👍

**Comment by lexiforest** [comment]
Does it make sense to set `CURLOPT_ACCEPT_ENCODING` based on the accept-encoding header? Will it fix your problem?

**Comment by ToshY** [comment]
> Does it make sense to set `CURLOPT_ACCEPT_ENCODING` based on the accept-encoding header? Will it fix your problem?

No, because the use-case is that when `Accept-Encoding` header is sent, the response should _not_ be compressed and instead let the PHP framework hande decompression (for better error handling). Setting the `CURLOPT_ACCEPT_ENCODING` based on `Accept-Encoding` header (e.g. `gzip`) would mean that it tries to decompress it, which is what I'm trying to prevent.  That's the main reason why I suggested an independant (environment variable) option to solve this.

**Comment by ToshY** [comment]
> I see. I think we should make a breaking change and just return the raw content from libcurl-impersonate. It's the way [how libcurl works](https://curl.se/mail/archive-2021-06/0031.html), anyway. The library user should then check the response header and decide if they want to decompress the content.

That's indeed a breaking one (which would be needed documented/communicated very well to avoid misunderstandings), but I think it's nice to have. By having the same behavior regaring decompression, it makes libcurl-impersonate and libcurl (more) interchangable.

> [!TIP]
> While I support the breaking change, if this decision is a bit too drastic, you could also choose to first create a **minor** release where the default behavior of automatic decompression still exists but an addtional (environment variable) option is supplied to disable decompression for those who need it. That way you don't create a breaking change right away and introduce the disable decompression option as an optional feature. Up to you 👍

**Comment by jingttkx** [comment]
HTTP_CONTENT_DECODING 

> > 明白了。我认为我们应该进行重大更改，并从 libcurl-impersonate 返回原始内容。无论如何，这就是 [libcurl 的工作](https://curl.se/mail/archive-2021-06/0031.html)方式。然后，库用户应检查响应标头并决定是否要解压缩内容。
> 
> 这确实是一个破坏性的（需要很好地记录/沟通以避免误解），但我认为拥有它很好。通过具有相同的行为来进行解压缩，它使 libcurl-impersonate 和 libcurl （更多） 可以互换。
> 
> 提示
> 
> 虽然我支持重大更改，但如果这个决定有点过于激烈，您也可以选择首先创建一个**次要**版本，其中自动解压缩的默认行为仍然存在，但提供了一个额外的（环境变量）选项来为需要它的人禁用解压缩。这样，您就不会立即创建重大更改，并引入禁用解压缩选项作为可选功能。轮到你了 👍

set CurlOpt.HTTP_CONTENT_DECODING=0 might be useful

`
from curl_cffi import requests, CurlOpt

response = requests.get("https://www.baidu.com", impersonate='edge', curl_options={
    CurlOpt.HTTP_CONTENT_DECODING:0 # default is 1
})
print(len(response.content))

`

**Comment by jingttkx** [comment]
> > I see. I think we should make a breaking change and just return the raw content from libcurl-impersonate. It's the way [how libcurl works](https://curl.se/mail/archive-2021-06/0031.html), anyway. The library user should then check the response header and decide if they want to decompress the content.
> 
> That's indeed a breaking one (which would be needed documented/communicated very well to avoid misunderstandings), but I think it's nice to have. By having the same behavior regaring decompression, it makes libcurl-impersonate and libcurl (more) interchangable.
> 
> Tip
> 
> While I support the breaking change, if this decision is a bit too drastic, you could also choose to first create a **minor** release where the default behavior of automatic decompression still exists but an addtional (environment variable) option is supplied to disable decompression for those who need it. That way you don't create a breaking change right away and introduce the disable decompression option as an optional feature. Up to you 👍

set CurlOpt.HTTP_CONTENT_DECODING=0 might be useful

```
from curl_cffi import requests, CurlOpt

response = requests.get("https://www.baidu.com", impersonate='edge', curl_options={
    CurlOpt.HTTP_CONTENT_DECODING:0 # default is 1
})
print(len(response.content))

```
